src/jdk.jfr/share/classes/jdk/jfr/internal/tool/Scrub.java

```diff
@@ -0,0 +1,194 @@
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.jfr.internal.tool;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import java.util.function.Predicate;
+
+import jdk.jfr.consumer.RecordedEvent;
+import jdk.jfr.consumer.RecordingFile;
+
+final class Scrub extends Command {
+
+    @Override
+    public String getName() {
+        return "scrub";
+    }
+
+    @Override
+    public List<String> getOptionSyntax() {
+        List<String> list = new ArrayList<>();
+        list.add("[--include-events <filter>]");
+        list.add("[--exclude-events <filter>]");
+        list.add("[--include-categories <filter>]");
+        list.add("[--exclude-categories <filter>]");
+        list.add("[--include-threads <filter>]");
+        list.add("[--exclude-threads <filter>]");
+        list.add("<input-file>");
+        list.add("[<output-file>]");
+        return list;
+    }
+
+    @Override
+    protected String getTitle() {
+        return "Scrub contents of a recording file";
+    }
+
+    @Override
+    public String getDescription() {
+        return getTitle() + ". See 'jfr help scrub' for details.";
+    }
+
+    @Override
+    public void displayOptionUsage(PrintStream stream) {
+        // 01234567890123456789012345678901234567890123467890123456789012345678901234567890
+        stream.println("  --include-events <filter>       Select events matching an event name");
+        stream.println();
+        stream.println("  --exclude-events <filter>       Exclude events matching an event name");
+        stream.println();
+        stream.println("  --include-categories <filter>   Select events matching a category name");
+        stream.println();
+        stream.println("  --exclude-categories <filter>   Exclude events matching a category name");
+        stream.println();
+        stream.println("  --include-threads <filter>      Select events matching a thread name");
+        stream.println();
+        stream.println("  --exclude-threads <filter>      Exclude events matching a thread name");
```

Comment:
Did you consider to allow filtering based on timestamps? It would be useful if one could extract a specific interval, e.g. 30 sec. starting at a given timestamp) from a larger recording.

Reply:
Yes, I also considered thresholds and random removal of events, but got lost in how the option syntax should look like. There is some precedence in JFR.dump that we might want to use, or not. In the end, I decided to keep the command line tool simple for now, all the filtering capabilities exists in the API. Most likely, I will revisit the tool later in the release to fix a few things. If time permits, I might look at time-based filtering as well then.

Reply:
That makes sense. Would you be open for a contribution by someone else (i.e. me) for the time-based filtering, once this initial work has been merged?

Reply:
Changes to command line options requires a CSR request and updating the documentation. I might as well do it when I fix other related issues, but it would be interesting to better understand the use case.

For JFR.dump there are four options, "begin", "end", "maxsize" and "maxage". Option "begin" and "end" accepts a timestamp but it can also be relative time, for example "begin=-30m", to dump the last 30 minutes, or you can do "maxage=30m".  It's possible to do combinations. For example, begin ="2022-02-23T12:49:54.913875Z" maxsize=100M to have a 100 MB before a particular timestamp. 

Not sure if that is what we want?

Reply:
I think "begin"/"end" with timestamps would suffice. One could provide both, so to get a slice from the middle of a recording, or just one, getting either the first part up to "end", or the last part, starting at "begin".